src/sulogin-shell/sulogin-shell.c

```diff
@@ -0,0 +1,73 @@
+/***
+  This file is part of systemd.
+
+  Copyright 2017 Felipe Sateler
+
+  systemd is free software; you can redistribute it and/or modify it
+  under the terms of the GNU Lesser General Public License as published by
+  the Free Software Foundation; either version 2.1 of the License, or
+  (at your option) any later version.
+
+  systemd is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <errno.h>
+#include <sys/prctl.h>
+
+#include "log.h"
+#include "process-util.h"
+#include "signal-util.h"
+
+static void fork_wait(const char* const cmdline[]) {
+        pid_t pid;
+
+        pid = fork();
+        if (pid < 0) {
+                log_error_errno(errno, "fork(): %m");
+                return;
+        }
+        if (pid == 0) {
+
+                /* Child */
+
+                (void) reset_all_signal_handlers();
+                (void) reset_signal_mask();
+                assert_se(prctl(PR_SET_PDEATHSIG, SIGTERM) == 0);
+
+                execv(cmdline[0], (char**) cmdline);
+                log_error_errno(errno, "Failed to execute %s: %m", cmdline[0]);
+                _exit(1); /* Operational error */
+        }
+
+        wait_for_terminate_and_warn(cmdline[0], pid, false);
+}
+
+static void print_mode(const char* mode) {
+        printf("You are in %s mode. After logging in, type \"journalctl -xb\" to view\n"
+                "system logs, \"systemctl reboot\" to reboot, \"systemctl default\" or ^D to boot\n"
+                "into default mode.\n", mode);
+        fflush(stdout);
+}
+
+int main(int argc, char *argv[]) {
+        static const char* const sulogin_cmdline[] = {SULOGIN, NULL};
+
+        log_set_target(LOG_TARGET_AUTO);
+        log_parse_environment();
+        log_open();
+
+        print_mode(argc > 1 ? argv[1] : "");
+
+        fork_wait(sulogin_cmdline);
+
+        execl(SYSTEMCTL_BINARY_PATH, "--no-block", "default", NULL);
```

Comment:
hmm, i'd prefer if we'd do the bus call here directly. It's relatively short, see fsck.c which does this (but use StartUnit() rather than StartUnitReplace())